contracts/governance/StakingRewards/StakingRewards.sol

```diff
@@ -0,0 +1,152 @@
+pragma solidity ^0.5.17;
+
+import "./StakingRewardsStorage.sol";
+import "../../openzeppelin/Initializable.sol";
+import "../../openzeppelin/SafeMath.sol";
+
+/**
+ * @title Staking Rewards contract.
+ * @notice This is a trial incentive program
+ * In this, the SOV emitted and becoming liquid from the Adoption Fund could be utilized
+ * to offset the higher APY's offered for Liquidity Mining events.
+ * Vesting contract stakes are excluded from these rewards.
+ * Only wallets which have staked previously liquid SOV are eligible for these rewards.
+ * Tokenholders who stake their SOV receive staking rewards, a pro-rata share
+ * of the revenue that the platform generates from various transaction fees
+ * plus revenues from stakers who have a portion of their SOV slashed for
+ * early unstaking.
+ * */
+contract StakingRewards is StakingRewardsStorage, Initializable {
+	using SafeMath for uint256;
+
+	/// @notice Emitted when SOV is withdrawn
+	/// @param receiver The address which recieves the SOV
+	/// @param amount The amount withdrawn from the Smart Contract
+	event SOVWithdrawn(address indexed receiver, uint256 amount);
+
+	/**
+	 * @notice Replace constructor with initialize function for Upgradable Contracts
+	 * This function will be called only once by the owner
+	 * */
+	function initialize(address _SOV, IStaking _staking) external onlyOwner initializer {
+		require(_SOV != address(0), "Invalid SOV Address.");
+		SOV = IERC20(_SOV);
+		staking = _staking;
+		startTime = block.timestamp;
+	}
+
+	/**
+	 * @notice Stops the current rewards program.
+	 * @dev All stakes existing on the contract at the point in time of
+	 * cancellation continue accruing rewards until the end of the staking
+	 * period being rewarded
+	 * */
+	function stop() external onlyOwner {
+		stopBlock = block.number;
+	}
+
+	/**
+	 * @notice Sets the base rate and divisor
+	 * @dev Base rate and divisor is used to calculate the rewards. The initial base
+	 * rate at the start of SIP-0024 is 29.75%. Base rate is annual
+	 * but we pay interest for 14 days, which is 1/26 of one staking year (1092 days)
+	 * @param _rate the base rate - it is the maximum interest rate(APY)
+	 * @param _divisor divisor is set as 2600000 = 26 (num periods per year) * 10 (max voting weight) * 10000 (2975 -> 0.2975)
+	 * @param _duration Max duration for which rewards can be collected at a go
```

Comment:
Updated the description.

Interesting finding - When I ran unit tests, the gasUsed by transaction for a scenario where the rewards are first claimed an year later came around 5M. But when I ran the coverage for the same unit test file, it showed approx 12M. Considering this, I believe we should set MaxDuration as 5-6 months initially. I am planning to initialise this in the deployment script. The maxDuration can be changed at any point in time by the owner.

gasUsed: 5057988
      ✓ should compute and send rewards to the staker a3 as applicable (2592ms)

gasUsed: 11848576
      ✓ should compute and send rewards to the staker a3 as applicable (18834ms)

provider.getGasPrice() will probably be useful when maxDuration needs to be changed on a case by case basis but I think that is not what we want.

Also, I checked for the possibility of adding a check in contract instead of looping through maxDuration, but that didn't seem feasible.

Reply:
The reason is that the tests coverage runs code without applying compiler optimization - if may have noticed that it recompiles contracts before running
I suggest leaving 1 year and test on the testnet watching the actual gas spent.